### ISSUE TITLE
automatic updates (or just checking for updates) for the browser and extensions

### ISSUE THREAD

**Issue by magicgoose** [Eloston/ungoogled-chromium#174]
This would a good thing, because outdated browsers and extensions are a security risk.
And checking manually every time is meh, I'm sure most users will just forget about it.

**Comment by Eloston** [comment]
We could probably create some sort of extension that users can install to have this functionality. Or we could have some external utility to check for updates.

I am open to suggestions.

**Comment by 9Morello** [comment]
An extension that checks for new Chromium versions would be cool. Not sure whats the best way for extensions, though.

**Comment by sp-1234** [comment]
https://blog.chromium.org/2009/07/smaller-is-faster-and-safer-too.html
https://dev.chromium.org/developers/design-documents/software-updates-courgette
tl;dr original chromium uses updates in form of binary diffs and they are very size-efficient

Probably if these diffs are generated and uploaded to "Github releases" too, they might work too.

**Comment by woolyss** [comment]
Henry updated its tool (chrlauncher) to install, update and launch ungoogled-chromium ;)
https://github.com/henrypp/chrlauncher/issues/33#issuecomment-279736691

...but nothing about extensions. Sorry.

**Comment by magicgoose** [comment]
@woolyss it's only for windows™.

**Comment by woolyss** [comment]
@magicgoose, Yes. It is only for Windows.

**Comment by Eloston** [comment]
It's a bit unnecessary to provide an external updater system for Linux, as most Linux systems come with a package system. For Mac, I think there are several packaging systems out there already.

**Comment by woolyss** [comment]
Yes @Eloston, I completely agree with you.

**Comment by koichirose** [comment]
Any news on this? Is there a way to at least have notifications about new updates?

**Comment by pjv** [comment]
i wrote a super-kludgy shell script with a [node package dependency](https://www.npmjs.com/package/chrome-web-store-item-property-cli), a dependency on [jq](https://stedolan.github.io/jq/), and a dependency on curl that looks at the extensions you have installed and then downloads the latest version from the play store into a directory of your choosing. You then have to drag them into the extensions page to install the updates.

It only works with play store extensions and it's a long way from a proper updater, but it beats doing it all manually, and I think if anyone wanted to it could be improved/extended.

Here's a [gist](https://gist.github.com/pjv/617f774dec75cc06dc7476f3ea21a3b5).

**Comment by Unkn0wn-MDCLXIV** [comment]
I've written a small script to update and install extensions easily.
It's windows only tho, but if there is demand I can upload and release it on github.
See my other reply at #226

**Comment by ajkblue** [comment]
I've found the easiest way to keep ungoogled-chromium updated is with brew cask, which will automatically install the latest version for me as long as I run "brew update".

**Comment by Eloston** [comment]
For users of feeds, an [Atom feed is available for the contributor binaries download page](https://raw.githubusercontent.com/ungoogled-software/ungoogled-chromium-binaries/master/feed.xml)

**Comment by sem5959** [comment]
Eloston - Give please the link to download your browser. DEBIAN

**Comment by antonin-lebrard** [comment]
@sem5959 https://ungoogled-software.github.io/ungoogled-chromium-binaries/

**Comment by sem5959** [comment]
Спасибо.

**Comment by unicorntaco** [comment]
I favor the idea of an extension to perform these various checks.  It would be great if that extension would accept SOCKS5 proxy (with auth) for googley chittering.

**Comment by ownchoice** [comment]
This is the only downside of using Ungoogled Chromium for me.
I don't want to have the browser and extensions outdated, and updating manually is a little bit painful.

**Comment by Eloston** [comment]
FYI my proposed solution for automatic extension updates is #285.

**Comment by intika** [comment]
**Update checker for UC :** 
Here are some extensions solution that can be used to check for update for UC other than a distro repo 

- Rss-Feed Reader like [feedbro](https://chrome.google.com/webstore/detail/feedbro/mefgmmbdailogpfhfblcnnjfmnpnmdfa?hl=en)

- Website Monitoring For Change Extensions : [visualping](https://chrome.google.com/webstore/detail/visualping/pemhgklkefakciniebenbfclihhmmfcd?hl=en) or [page-monitor](https://chrome.google.com/webstore/detail/page-monitor/ogeebjpdeabhncjpfhgdibjajcajepgg?hl=en) 
You just need add [uc-site](https://ungoogled-software.github.io/ungoogled-chromium-binaries/) to the monitored sites... 

- Github notification [notifier-for-github](https://chrome.google.com/webstore/detail/notifier-for-github/lmjdlojahmbbcodnpecnjnmlddbkjhnn) (does not require to be logged in)

**Comment by Eloston** [comment]
In addition to the Atom feed for contributor binaries ([link above](https://github.com/Eloston/ungoogled-chromium/issues/174#issuecomment-372207201)), there's also [the Atom feed for new tags from GitHub](https://github.com/Eloston/ungoogled-chromium/tags.atom).

**Comment by intika** [comment]
a function chat check that link could be added easily mainstream but in one hand 

- update are important for security 
and the other
- users may dislike a connection to github

I think as it would just check one link for updates i think it would be fair to enable it by default (this eventual feature) and add a flag to make it disableable 

@Eloston what do you think ?

**Comment by Eloston** [comment]
@intika This is something an extension can do perfectly well. I see no need for anything to be added to ungoogld-chromium for this. Once a solution comes around, we could suggest it in the Wiki.

**Comment by IANatCAMBIO** [comment]
I agree in leaving this functionality as an extension. The modular framework is there for a reason, and should be used whenever possible. 

It sounds like this is really 2 issues:
Checking/Notifying that a new version is available
and
Performing the update.

BSD/Linux and MacOS can use their package managers for #2, but a notification of a new version would be nice.

Windows users would need both types of functionality.

As Intika said above, extensions to check for page changes (i.e. new published binaries) already exist, but they must be setup and configured. If we're trying to stay true to the published purpose for this project, then we try to maintain functional parity with Chrome/Chromium. That would include at least notifying of updates.

Would it be possible to have a bundle of default extensions (including and update checker) to provide the functionality? This does not have to be included with UC, so that we still maintain the 'you must opt-in for new features' requirement.
Although, it should not be enabled by default, branding and hosting that bundle along with the binary downloads would allow for standard Open-Source protections (code review).

**Comment by NeverDecaf** [comment]
I threw together a barebones extension that adds this functionality to Ungoogled Chromium:
https://github.com/NeverDecaf/chromium-web-store

Unfortunately I could not find a way with my limited javascript knowledge to dynamically detect the buttons on the chrome web store, therefore if the site layout changes at all this part of the extension will surely break. The code for update checking, however, uses the chrome api and should work fine until google deprecates it.  
  
Please note I have barely tested this, and on Windows only. If something is broken I'll be happy to fix it if I can, just open an issue. A PR to add a robust button detection or to make the popup actually look nice is more than welcome.

**Comment by woolyss** [comment]
Thank you very much NeverDecaf. It works like a charm. Bravo!

I made a video to show it. Free to use and share it anywhere!
- Stored on my server: https://chromium.woolyss.com/f/video-extension-chromium-web-store.mp4
- Stored at Streamable (without any user account):  https://streamable.com/655nn

My config:
- ungoogled-chromium 76.0.3809.100 (64-bit)
- chrlaucher (https://github.com/henrypp/chrlauncher) to test ungoogled-chromium in its portable version.
- Windows 10 1903 with the enabled dark theme.

**Comment by NeverDecaf** [comment]
@woolyss Thanks for the video! I've added it to the readme as it makes the install process very clear.

**Comment by woolyss** [comment]
It is ok @NeverDecaf. Thank you. I put a link of your github project on my site. Contact me if you want a better video! ;)

**Comment by TNTBOMBOM** [comment]
Appimage support update, any issue with it? This resolves almost all GNU/Linux upgrades issue.

https://github.com/AppImage/AppImageUpdate

**Comment by anoxi** [comment]
How can i automatically upgrade the ungoogled-chromium package on Ubuntu using unattended-upgrades?
The following apt policy from the opensuse repository does not seem to be not compatible with the Unattended-Upgrade::Allowed-Origins pattern `"<o>:<a>"` ` (<origin>:<archive>) `

```
$ apt-cache policy
 500 http://download.opensuse.org/repositories/home:/ungoogled_chromium/Ubuntu_Focal  Packages
     release o=obs://build.opensuse.org/home:ungoogled_chromium/Ubuntu_Focal,n=Ubuntu_Focal,l=home:ungoogled_chromium,c=
     origin download.opensuse.org
```
The policy is missing the archive tag and the origin tag format including ":"(obs://build.opensuse.org/home:ungoogled_chromium/Ubuntu_Focal) is incompatible.

**Comment by PF4Public** [comment]
> outdated […] extensions are a security risk.

In the light of a recent news about an extension being sold to people, whose intentions were not very decent, auto-updating mechanism of Chrome backfired badly on users. So that auto-updated extension also poses a security risk!

**Comment by TNTBOMBOM** [comment]
can you share with us the sources? "extension being sold to people" , 
"chrome backfired badly on users"

ThX

PF4Public:
>> outdated […] extensions are a security risk.
> 
> In the light of a recent news about an extension being sold to people, whose intentions were not very decent, auto-updating mechanism of Chrome backfired badly on users. So that auto-updated extension also poses a security risk!
>

**Comment by PF4Public** [comment]
Given the tremendous hatred that the author of the extension received, I hesitate providing direct links, as this might spawn another wave of anger, which is unconstrictive and absolutely not helpful in this situation.

I think, I'll just describe the security issue itself instead.
Imagine an extension, that is installed on a lot of Chrome profiles. One day, author of the extension gives access to posting updates to someone else. That person modifies the extension to send headers of a select requests of victims to himself. Due to autoupdate, this malicious extension got installed on a computers of unsuspecting users without their approval or consent. 

Aftermath is still unfolding unfortunately :(

**Comment by dm17** [comment]
@woolyss @NeverDecaf You guys think it would be easier to just have a Dockerfile with a config file sources repos for each extension we want - and we just rebuild this every time we want to latest plugins & ungoogled-chromium? Thoughts?
I also have different extensions per chromium profile, which could be a pain for this.

**Comment by selurvedu** [comment]
> For users of feeds, an [Atom feed is available for the contributor binaries download page](https://raw.githubusercontent.com/ungoogled-software/ungoogled-chromium-binaries/master/feed.xml)

@Eloston this is great! Would it b possible to create OS/distribution-specific feeds?

**Comment by Jon-guy30** [comment]
Hi! 
I've just started using ungoogled-chromium and wondering about the auto-update issue. Currently I'm subscribed to the ungoogled Atom feed linked above. I think the updating of this browser could be made easier, for example I really like how LibreWolf handles it, they have an unofficial LibreWolf updater extension. I think ungoogled-chromium could really benefit from such an extension. We already have a Google WebStore updater, now it just needs a browser update handler. 
While Atom feeds are good, it's not really good for non-technical people. They want easy, hassle-free updating, or at least an update notification within the web browser software. 

I've also looked at chrlauncher. I'm not sure how to configure it to be just an updater, from what I've read it downloads an entire chromium browser to the computer, I don't want another browser installed, I want it to update only.
These are my chrlauncher settings, at least the important ones:

```
ChromiumDirectory=C:\Users\User\AppData\Local\Chromium\Application
ChromiumUpdateUrl=https://github.com/ungoogled-software/ungoogled-chromium/tags (uncommented, I obviously put the GitHub tags URL in it)
# Chromium executable file name (string):
ChromiumBinary=chrome.exe
# Use chrlauncher as updater, but does not start Chromium (boolean):
#
# false	-> update & start Chromium (default)
# true	-> download & install Chromium update without start
ChromiumUpdateOnly=true
ChromiumType=ungoogled-chromium

```
Could anyone tell me if this is how it's done or is it incorrect? Thanks! The documentation of chrlauncher isn't really helping me.

**Comment by axcairns** [comment]
> > For users of feeds, an [Atom feed is available for the contributor binaries download page](https://raw.githubusercontent.com/ungoogled-software/ungoogled-chromium-binaries/master/feed.xml)
> 
> @Eloston this is great! Would it b possible to create OS/distribution-specific feeds?

https://github.com/ungoogled-software/ungoogled-chromium-windows/releases.atom

**Comment by dm17** [comment]
> Hi! I've just started using ungoogled-chromium and wondering about the auto-update issue. Currently I'm subscribed to the ungoogled Atom feed linked above. I think the updating of this browser could be made easier, for example I really like how LibreWolf handles it, they have an unofficial LibreWolf updater extension. I think ungoogled-chromium could really benefit from such an extension. We already have a Google WebStore updater, now it just needs a browser update handler. While Atom feeds are good, it's not really good for non-technical people. They want easy, hassle-free updating, or at least an update notification within the web browser software.
> 
> I've also looked at chrlauncher. I'm not sure how to configure it to be just an updater, from what I've read it downloads an entire chromium browser to the computer, I don't want another browser installed, I want it to update only. These are my chrlauncher settings, at least the important ones:
> 
> ```
> ChromiumDirectory=C:\Users\User\AppData\Local\Chromium\Application
> ChromiumUpdateUrl=https://github.com/ungoogled-software/ungoogled-chromium/tags (uncommented, I obviously put the GitHub tags URL in it)
> # Chromium executable file name (string):
> ChromiumBinary=chrome.exe
> # Use chrlauncher as updater, but does not start Chromium (boolean):
> #
> # false	-> update & start Chromium (default)
> # true	-> download & install Chromium update without start
> ChromiumUpdateOnly=true
> ChromiumType=ungoogled-chromium
> ```
> 
> Could anyone tell me if this is how it's done or is it incorrect? Thanks! The documentation of chrlauncher isn't really helping me.

Figured that out yet?

**Comment by claudiodekker** [comment]
Since this appears to be a common and recurring issue (arguably for good reason), I figured I'd try to partially address it by creating **a Chromium extension that automatically checks for new Chromium updates in-line with your preferences**.

Its functionality resembles that of NeverDecaf's "Chromium Web Store" extension: It displays a badge indicating available updates and, when interacted with, provides a concise overview of all available updates, allowing you to directly download newer build versions, jump to the release page, copy hashes and more. Essentially, it serves as the closest thing to an automated updater mechanism, without one being built-in.

You can find it here: https://github.com/claudiodekker/chromium-update-notifier
